README.md
# @mapcat/mapview-init

The *@mapcat/mapview-init* javascript library is used to obtain MAPCAT raster tile url and vector tile style sheet, that you can use for your [Leaflet](http://leafletjs.com/), [OpenLayers](https://openlayers.org/) and [Tangram](https://mapzen.com/products/tangram/) projects. You can find the prebuilt version of the javascript library under `./dist` folder.

## Build dependencies
* node (min _v6.9.5_)
* npm (min _4.3.0_)
* yarn (min _1.3.2_)

## Build
* Clone or download zip file from [MAPCATcom/mapcat-mapview-init](https://github.com/MAPCATcom/mapcat-mapview-init.git) repository  
`git clone https://github.com/MAPCATcom/mapcat-mapview-init.git`
* Go to the cloned (or unzipped) project main folder  
`cd mapcat-mapview-init`
* Install Node.js dependencies  
`yarn` or `yarn install`
* To build release version  
`yarn build-min`
* To build debug version  
`yarn build-dev`

If the build was successful you can find under `./dist` folder the recently built release (`mapcatview-min.js`) or debug (`mapcatview-dev.js`) version of library.  

## Usage
Create your HTML skeleton and link mapcat-mapview-init library before your own custom script file.  
`<script src="mapcatview-min.js"></script>`

### In your script file:  

To initialize Mapcat mapview use one of the following functions  
* for raster tiles:  
   
`mapcatview.initRasterView(callback, apiKey, layerOptions, rasterOptions);`

* for vector tiles:  

`mapcatview.initVectorView(callback, apiKey, layerOptions, vectorOptions);`
  
#### Example

```javascript
mapcatview.initRasterView(function(error, response) {

    //your code

}, '< Your MAPCAT Visualization API key >');

mapcatview.initVectorView(function(error, response) {

    //your code

}, '< Your MAPCAT Visualization API key >');
```

### Functions parameters

#### callback (error, response): *function* `(required)` - Callback function  
It gets called when the map initialization request returns from our server. The first parameter holds the error message (`null` means no error), the second parameter in case of initRasterView function call is the response data holding your templated map view url (*string*). In case of using initVectorView the response data holds the vector tile style sheet (*object*).

#### apiKey: *string* `(required)` - Your MAPCAT Visualization API key

#### layerOptions: *object* `(optional)` - Options to show cycle roads, routes layers  
Layers are used to toggle specific subsets of data rendered on the raster and vector tiles. Customizable: cycle roads and routes. Default: cycle road and route layers are off.
##### Example
```javascript
{
    cycle: {
        road: true,
        route: false
    }
}
```

#### vectorOptions: *object* `(optional)` - Options to change style sheet format
Customizable: vector style sheet format returned, possible values: `"mapbox"`, `"openlayers"`, `"tangram"`. This defaults to: `"mapbox"`.
##### Example
```javascript
{
    styleSheet: "mapbox"
}
```

#### rasterOptions: *object* `(optional)` - Options to set label language and raster tile scale  
Parameter *lang* is the ISO 639-1 language code representation of the desired language, it defaults to `"en"`. To disable label rendering set lang to `null` or `""`.  
Parameter *scale* must be `1` or `2`. By default, it is `1`, meaning that the required raster tile size is 256 × 256 pixel. If you want to get tiles for retina displays (512 × 512 pixel tiles), you can use value `2`.  
##### Example
```javascript
{
    lang: "hu",
    scale: 1
}
```

### Example with [Leaflet JS](http://leafletjs.com/)
In your HTML page `<head>` include Leaflet CSS file
```html
<link rel="stylesheet" href="https://unpkg.com/leaflet@1.2.0/dist/leaflet.css" />
```
Put a div element in `<body>` section to your map
```html
<div id="map" style="width: 400px; height: 300px;"></div>
```
Include Leaflet JavaScript, mapcat-mapview-init.js and your own script file in this order
```html
<script src="https://unpkg.com/leaflet@1.2.0/dist/leaflet.js"></script>
<script src="mapcatview-min.js"></script>
<script src="script.js"></script>
```
In your script file initialize your map
```javascript
var layerOptions = {
    cycle: {
        road: true,
        route: true
    }
};

var rasterOptions = {
    lang: 'en',
    scale: 1
};

mapcatview.initRasterView(function(error, response) {
    if (error) {
        
        // error handling

        console.log(error);
        return;
    }

    var tileUrl = response;

    var map = L.map('map', {
        zoomControl: false,
        center: L.latLng(47.4979, 19.0402),
        zoom: 13,
        minZoom: 0,
        maxZoom: 18
    });

    L.tileLayer(tileUrl, {
        attribution: 'Imagery &copy; 2017 <a href="https://mapcat.com">MAPCAT</a>, Map data &copy; <a href="http://osm.org/copyright">OpenStreetMap</a contributors',
        maxZoom: 18
    }).addTo(map);

    // your code

}, '< Your MAPCAT Visualization API key >', layerOptions, rasterOptions);
```

### Example with [Mapbox GL JS](https://github.com/mapbox/mapbox-gl-js)
In your HTML page `<head>` include Mapbox GL CSS file
```html
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/mapbox-gl/0.43.0/mapbox-gl.css" />
```
Put a div element in `<body>` section to your map
```html
<div id="map" style="width: 400px; height: 300px;"></div>
```
Include Mapbox GL JS JavaScript, mapcat-mapview-init.js and your own script file in this order
```html
<script src="https://cdnjs.cloudflare.com/ajax/libs/mapbox-gl/0.43.0/mapbox-gl.js"></script>
<script src="mapcatview-min.js"></script>
<script src="script.js"></script>
```
In your script file initialize your map
```javascript
var layerOptions = {
    cycle: {
        road: true,
        route: true
    }
};

var vectorOptions = {
    styleSheet: "mapbox"
};

mapcatview.initVectorView(function(error, response) {
    if (error) {

        // error handling

        console.log(error);
        return;
    }

    var styleSheet = response;

    var map = new mapboxgl.Map({
        container: 'map',
        style: styleSheet,
        center: [0, 51.5],
        zoom: 13
    });

    // your code

}, '< Your MAPCAT Visualization API key >', layerOptions, vectorOptions);
```

Substitute `< Your MAPCAT Visualization API key >` with your Visualization API key.  

Read [MAPCAT for Developers](https://docs.mapcat.com) for more information and examples.

## Licence
[MIT Licence](https://github.com/MAPCATcom/mapcat-mapview-init/blob/master/LICENSE)


CHANGELOG.md
## 0.0.9 (May 4, 2018)
- Removed style type validation from client side

## 0.0.8 (April 18, 2018)
- Handle Tangram compatible style sheet

## 0.0.7 (February 21, 2018)
- Access token renamed to API key

## 0.0.6 (January 18, 2018)
- Fix: https protocol handling

## 0.0.5 (January 17, 2018)
- Added separate vector style sheet option parameter

## 0.0.4 (January 9, 2018)
- Follow MapInit API url changes
- Vector map style is returned from MapInit API call
- Handle OpenLayers and Mapbox compatible style sheets
- Autodetect usable url protocol

## 0.0.3 (January 3, 2018)
- Change API url

## 0.0.2 (November 27, 2017)
- Forward server error response
- Response url protocol fix

## 0.0.1 (November 24, 2017)
- First release


src/mapcat-mapview-init.js
/*eslint no-undef: "error"*/
/*eslint-env node*/
var https = require('https');

function initRasterView (callback, apiKey, layerOptions, rasterOptions) {
    var type = 'raster';
    initView(callback, type, apiKey, layerOptions, null, rasterOptions);
}

function initVectorView (callback, apiKey, layerOptions, vectorOptions) {
    var type = 'vector';
    initView(callback, type, apiKey, layerOptions, vectorOptions, null);
}

function initView (callback, type, apiKey, layerOptions, vectorOptions, rasterOptions) {
    var data = {};
    data.protocol = 'http';
    if (document && document.location && document.location.protocol && document.location.protocol === 'https:') {
        data.protocol = 'https';
    }
    try {
        if (apiKey === undefined || apiKey === null || typeof(apiKey) !== 'string' || apiKey.length === 0) {
            throw new Error('Invalid MAPCAT Visualization API key! Expected: string value');
        }

        // parse layer options
        var layers = {
            base: '',
            ocean: '',
            relief: '',
            landcover: ''
        };
        if (layerOptions && layerOptions.cycle) {
            if (layerOptions.cycle.road && typeof(layerOptions.cycle.road) !== 'boolean') {
                throw new Error('Layer option cycle.road set, but invalid type. Expected: true or false');
            }
            if (layerOptions.cycle.route && typeof(layerOptions.cycle.route) !== 'boolean') {
                throw new Error('Layer option cycle.route set, but invalid type. Expected: true or false');
            }
            if (layerOptions.cycle.road === true && layerOptions.cycle.route === true) {
                layers.cycle = '';
            } else if (layerOptions.cycle.road === true) {
                layers.cycle = '--,road';
            } else if (layerOptions.cycle.route === true) {
                layers.cycle = '--,route';
            }
        }
        data.layers = layers;

        if (type === 'vector') {
            // parse vector options
            if (vectorOptions && vectorOptions.styleSheet) {
                data.type = vectorOptions.styleSheet;
            }
        } else if (type === 'raster') {
            // parse raster options
            if (rasterOptions && rasterOptions.lang !== undefined && rasterOptions.lang !== null) {
                if (typeof(rasterOptions.lang) !== 'string') {
                    throw new Error('Invalid language parameter type! Expected: string value or null');
                }
                data.lang = rasterOptions.lang;
            } else if (rasterOptions && rasterOptions.lang === null) {
                data.lang = '';
            } else {
                data.lang = 'en';
            }
            if (rasterOptions && rasterOptions.scale !== undefined && rasterOptions.scale !== null) {
                var s = Number(rasterOptions.scale);
                if (s === 1 || s === 2) {
                    data.scale = s;
                } else {
                    throw new Error('Invalid scale! Expected scale value: 1 or 2');
                }
            } else {
                data.scale = 1;
            }
        }
    } catch (error) {
        return callback(error.message);
    }

    var postData = JSON.stringify(data);
    var options = {
        host: 'api.mapcat.com',
        port: 443,
        path: '/api/mapinit/' + type,
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            'Content-Length': Buffer.byteLength(postData),
            'X-Api-Key': apiKey
        }
    };

    var cb = function (response) {
        var str = '';
        response.on('data', function (chunk) {
            str += chunk;
        });

        response.on('end', function () {
            var url = '';
            var resJson = {};
            try {
                resJson = JSON.parse(str);
                if (resJson.error) {
                    return callback(resJson.error);
                } else if (type === 'raster' && resJson.url && typeof(resJson.url) === 'string') {
                    url = resJson.url;
                } else if (type === 'vector') {
                    return callback(null, resJson);
                } else {
                    return callback('Unexpected error! ' + str);
                }
            } catch (error) {
                return callback('Unable to parse response! Error: ' + str);
            }
            callback(null, url);
        });
    };
    var req = https.request(options, cb);
    req.write(postData);
    req.end();
    req.on('error', function (err) {
        return callback(err.message);
    });
}

module.exports.initRasterView = initRasterView;
module.exports.initVectorView = initVectorView;


example/script.js
var map;

function init(){
    map = L.map('map', {
        zoomControl: false,
        center: L.latLng(47.4979, 19.0402),
        zoom: 13,
        minZoom: 0,
        maxZoom: 18
    });
}

function load(layerOptions, rasterOptions){
    mapcatview.initRasterView(function(error, response) {
        if (error) {
            console.log(error);
            return;
            // Error handling...
        }
        var tileUrl = response;
        L.tileLayer(tileUrl, {
            attribution: 'Imagery &copy; 2017 <a href="https://mapcat.com">MAPCAT</a>, Map data &copy; <a href="http://osm.org/copyright">OpenStreetMap</a contributors',
            maxZoom: 18
        }).addTo(map);
    }, 'jM9oGlsfWxOOYYF0kvuq2UbYl3XrVuUzJmwfnB6M', layerOptions, rasterOptions);
}

function reload(){
    var layerOptions = null;
    var rasterOptions = null;
    try {
        layerOptions = JSON.parse(document.getElementById('layerOp').value);
    } catch (e) {
        console.log('Error:', e);
        alert('Unable to parse layer options');
    }
    try {
        rasterOptions = JSON.parse(document.getElementById('rasterOp').value);
    } catch (e) {
        console.log('Error:', e);
        alert('Unable to parse raster options');
    }
    load(layerOptions, rasterOptions);
}

init();
reload();